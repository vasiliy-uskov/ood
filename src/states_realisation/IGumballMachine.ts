export interface IGumballMachine {
	releaseBall(): void;

	getBallCount(): number;

	setSoldOutState(): void;

	setNoQuarterState(): void;

	setHasQuarterState(): void;
}
