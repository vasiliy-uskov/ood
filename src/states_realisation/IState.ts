export interface IState {
	insertQuarter(): void;

	ejectQuarter(): void;

	turnCrank(): void;

	dispense(): void;

	toString(): string;
}