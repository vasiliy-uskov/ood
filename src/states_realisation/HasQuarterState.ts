import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {ILogger} from "../ILogger";

export class HasQuarterState implements IState {
	private readonly _gumballMachine: IGumballMachine;
	private readonly _logger: ILogger;

	constructor(gumballMachine: IGumballMachine, logger: ILogger) {
		this._gumballMachine = gumballMachine;
		this._logger = logger;
	}

	insertQuarter(): void {
		this._logger.log("You can't insert another quarter");
	}

	ejectQuarter(): void {
		this._logger.log("Quarter returned");
		this._gumballMachine.setNoQuarterState();
	}

	turnCrank(): void {
		this._logger.log("You turned...");
		this._gumballMachine.setSoldState();
	}

	dispense(): void {
		this._logger.log("No gumball dispensed");
	}

	toString(): string {
		return "waiting for turn of crank";
	}
}