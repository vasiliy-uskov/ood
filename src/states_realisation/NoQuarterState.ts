import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {ILogger} from "../ILogger";

export class NoQuarterState implements IState {
	private readonly _gumballMachine: IGumballMachine;
	private readonly _logger: ILogger;

	constructor(gumballMachine: IGumballMachine, logger: ILogger) {
		this._gumballMachine = gumballMachine;
		this._logger = logger;
	}

	insertQuarter(): void {
		this._logger.log("You inserted a quarter");
		this._gumballMachine.setHasQuarterState();
	}

	ejectQuarter(): void {
		this._logger.log("You haven't inserted a quarter");
	}

	turnCrank(): void {
		this._logger.log("You turned but there's no quarter");
	}

	dispense(): void {
		this._logger.log("You need to pay first");
	}

	toString(): string {
		return "waiting for quarter";
	}
}