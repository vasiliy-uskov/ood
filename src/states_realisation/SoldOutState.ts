import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {ILogger} from "../ILogger";

export class SoldOutState implements IState {
	private readonly _gumballMachine: IGumballMachine;
	private readonly _logger: ILogger;

	constructor(gumballMachine: IGumballMachine, logger: ILogger) {
		this._gumballMachine = gumballMachine;
		this._logger = logger;
	}

	insertQuarter(): void {
		this._logger.log("You can't insert a quarter, the machine is sold out");
	}

	ejectQuarter(): void {
		this._logger.log("You can't eject, you haven't inserted a quarter yet");
	}

	turnCrank(): void {
		this._logger.log("You turned but there's no gumballs");
	}

	toString(): string {
		return "sold out";
	}
}