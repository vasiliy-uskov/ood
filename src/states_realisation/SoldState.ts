import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {ILogger} from "../ILogger";

export class SoldState implements IState {
	private readonly _gumballMachine: IGumballMachine;
	private readonly _logger: ILogger;

	constructor(gumballMachine: IGumballMachine, logger: ILogger) {
		this._gumballMachine = gumballMachine;
		this._logger = logger;
	}

	insertQuarter(): void {
		this._logger.log("Please wait, we're already giving you a gumball");
	}

	ejectQuarter(): void {
		this._logger.log("Sorry you already turned the crank");
	}

	turnCrank(): void {
		this._logger.log("Turning twice doesn't get you another gumball");
	}

	dispense(): void {
		this._gumballMachine.releaseBall();
		if (this._gumballMachine.getBallCount() == 0) {
			this._logger.log("Oops, out of gumballs");
			this._gumballMachine.setSoldOutState();
		}
		else {
			this._gumballMachine.setNoQuarterState();
		}
	}

	toString(): string {
		return "delivering a gumball";
	}
}