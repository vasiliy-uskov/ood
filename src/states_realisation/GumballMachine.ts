import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {SoldOutState} from "./SoldOutState";
import {NoQuarterState} from "./NoQuarterState";
import {HasQuarterState} from "./HasQuarterState";
import {ILogger} from "../ILogger";

export class GumballMachine {
	private readonly _logger: ILogger;
	private readonly _soldOutState: SoldOutState;
	private readonly _noQuarterState: NoQuarterState;
	private readonly _hasQuarterState: HasQuarterState;
	private _state: IState;
	private _count = 0;

	constructor(numBalls: number, logger: ILogger) {
		this._logger = logger;
		this._count = numBalls;
		const interfaceImpl = this._getGumballMachineInterface();
		this._soldOutState = new SoldOutState(interfaceImpl, logger);
		this._noQuarterState = new NoQuarterState(interfaceImpl, logger);
		this._hasQuarterState = new HasQuarterState(interfaceImpl, logger);
		if (this._count > 0) {
			this._state = this._noQuarterState;
		}
		else {
			this._state = this._soldOutState;
		}
	}

	ejectQuarter() {
		this._state.ejectQuarter();
	}

	insertQuarter() {
		this._state.insertQuarter();
	}

	turnCrank() {
		this._state.turnCrank();
	}

	refill(count: number) {
		this._count = count;
		if (count > 0) {
			this._setNoQuarterState();
		}
		else {
			this._setSoldOutState();
		}
	}

	toString(): string {
		return `
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: ${this._count} gumball${this._count != 1 ? "s" : ""}
Machine is ${this._state.toString()}
`;
	}

	private _getGumballMachineInterface(): IGumballMachine {
		return {
			getBallCount: this._getBallCount.bind(this),
			releaseBall: this._releaseBall.bind(this),
			setSoldOutState: this._setSoldOutState.bind(this),
			setNoQuarterState: this._setNoQuarterState.bind(this),
			setHasQuarterState: this._setHasQuarterState.bind(this),
		}
	}

	private _getBallCount(): number {
		return this._count;
	}

	private _releaseBall(): void {
		if (this._count != 0) {
			this._logger.log("A gumball comes rolling out the slot...");
			--this._count;
		}
	}

	private _setSoldOutState(): void {
		this._state = this._soldOutState;
	}

	private _setNoQuarterState(): void {
		this._state = this._noQuarterState;
	}

	private _setHasQuarterState(): void {
		this._state = this._hasQuarterState;
	}
}