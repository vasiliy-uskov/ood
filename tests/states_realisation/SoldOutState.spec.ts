import {SoldOutState} from "../../src/states_realisation/SoldOutState";
import {MockGumballMachine} from "./MockGumballMachine";
import {MockLogger} from "../MockLogger";


it('toString', () => {
	const state = new SoldOutState(new MockGumballMachine(), new MockLogger());
	expect(state.toString()).toMatchSnapshot();
});

it('insertQuarter', () => {
	const gumballMachine = new MockGumballMachine();
	const state = new SoldOutState(gumballMachine, new MockLogger());
	state.insertQuarter();
	gumballMachine.checkGumballMachineDoNotChangeState();
});

it('ejectQuarter', () => {
	const gumballMachine = new MockGumballMachine();
	const state = new SoldOutState(gumballMachine, new MockLogger());
	state.ejectQuarter();
	gumballMachine.checkGumballMachineDoNotChangeState();
});

it('turnCrank', () => {
	const gumballMachine = new MockGumballMachine();
	const state = new SoldOutState(gumballMachine, new MockLogger());
	state.turnCrank();
	gumballMachine.checkGumballMachineDoNotChangeState();
});