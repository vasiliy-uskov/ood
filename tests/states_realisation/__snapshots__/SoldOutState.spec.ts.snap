// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toString 1`] = `"sold out"`;
